File: app/src/main/java/com/github/android/lvrn/lvrnproject/service/core/TaskService.java
package com.github.android.lvrn.lvrnproject.service.core;

import android.support.annotation.NonNull;

import com.github.valhallalabs.laverna.persistent.entity.Task;
import com.github.android.lvrn.lvrnproject.service.ProfileDependedService;
import com.github.android.lvrn.lvrnproject.service.form.TaskForm;
import com.github.android.lvrn.lvrnproject.util.PaginationArgs;

import java.util.List;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

public interface TaskService extends ProfileDependedService<Task, TaskForm> {

    /**
     * A method which retrieves an amount of uncompleted tasks from a start position by a profile
     * id.
     * @param profileId an id of a profile.
     * @param paginationArgs arguments of pagination such as offset and limit.
     * @return a list of entities.
     */
    @NonNull
    List<Task> getUncompletedByProfile(@NonNull String profileId, @NonNull PaginationArgs paginationArgs);


    /**
     * A method which retrieves an amount of uncompleted tasks from a start position by a profile
     * id and a description
     * @param profileId an id of a profile.
     * @param description a description of the task.
     * @param paginationArgs arguments of pagination such as offset and limit.
     * @return a list of entities.
     */
    @NonNull
    List<Task> getUncompletedByProfileAndDescription(@NonNull String profileId, @NonNull String description, @NonNull PaginationArgs paginationArgs);

    /**
     * A method which retrieves entities by a note id.
     * @param noteId an id of note.
     * @return a list of entities.
     */
    @NonNull
    List<Task> getByNote(@NonNull String noteId);
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/service/core/TaskService.kt
package com.github.android.lvrn.lvrnproject.service.core

import com.github.android.lvrn.lvrnproject.service.ProfileDependedService
import com.github.android.lvrn.lvrnproject.service.form.TaskForm
import com.github.android.lvrn.lvrnproject.util.PaginationArgs
import com.github.valhallalabs.laverna.persistent.entity.Task

interface TaskService : ProfileDependedService<Task, TaskForm> {

    /**
     * A method which retrieves an amount of uncompleted tasks from a start position by a profile
     * id.
     * @param profileId an id of a profile.
     * @param paginationArgs arguments of pagination such as offset and limit.
     * @return a list of entities.
     */
    fun getUncompletedByProfile(profileId: String, paginationArgs: PaginationArgs): List<Task>


    /**
     * A method which retrieves an amount of uncompleted tasks from a start position by a profile
     * id and a description
     * @param profileId an id of a profile.
     * @param description a description of the task.
     * @param paginationArgs arguments of pagination such as offset and limit.
     * @return a list of entities.
     */
    fun getUncompletedByProfileAndDescription(
        profileId: String,
        description: String,
        paginationArgs: PaginationArgs,
    ): List<Task>

    /**
     * A method which retrieves entities by a note id.
     * @param noteId an id of note.
     * @return a list of entities.
     */
    fun getByNote(noteId: String): List<Task>
}